File: Graphs/bfs_graph.py
# File Name: bfs_graph.py

from collections import deque

def bfs(start, visited, graph):
    queue = deque()
    visited[start] = True
    queue.append(start)

    while queue:
        current = queue.popleft()
        print(current, end=" ")

        for neighbor in graph[current]:
            if not visited[neighbor]:
                visited[neighbor] = True
                queue.append(neighbor)

vertices = int(input("Enter number of vertices: "))
edges = int(input("Enter number of edges: "))

graph = [[] for _ in range(vertices)]
visited = [False] * vertices

print("Enter edges (u v):")
for _ in range(edges):
    u, v = map(int, input().split())
    graph[u].append(v)
    graph[v].append(u)  # Undirected graph

start = int(input("Enter starting vertex for BFS: "))

print("BFS Traversal:", end=" ")
bfs(start, visited, graph)

File: Graphs/BFSGraph.java
// File Name: BFSGraph.java

import java.util.*;

public class BFSGraph {

    static ArrayList<ArrayList<Integer>> graph = new ArrayList<>();
    static boolean[] visited;

    static void bfs(int start) {
        Queue<Integer> queue = new LinkedList<>();

        visited[start] = true;
        queue.add(start);

        while (!queue.isEmpty()) {
            int current = queue.poll();
            System.out.print(current + " ");

            for (int neighbor : graph.get(current)) {
                if (!visited[neighbor]) {
                    visited[neighbor] = true;
                    queue.add(neighbor);
                }
            }
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.print("Enter number of vertices: ");
        int vertices = sc.nextInt();

        System.out.print("Enter number of edges: ");
        int edges = sc.nextInt();

        visited = new boolean[vertices];

        for (int i = 0; i < vertices; i++)
            graph.add(new ArrayList<>());

        System.out.println("Enter edges (u v):");
        for (int i = 0; i < edges; i++) {
            int u = sc.nextInt();
            int v = sc.nextInt();
            graph.get(u).add(v);
            graph.get(v).add(u);  // Undirected graph
        }

        System.out.print("Enter starting vertex for BFS: ");
        int start = sc.nextInt();

        System.out.print("BFS Traversal: ");
        bfs(start);
    }
}
